File: backend-express/auth_utils.js
const crypto = require('crypto');


const SALT_LENGTH = 16;
const PBKDF2_ITERATIONS = 500000;
const PASSWORD_KEY_LENGTH = 64;
const PBKDF2_DIGEST_ALGO = 'sha256';

function _hex_from_bytes(buffer) {
    return [...new Uint8Array(buffer)]
        .map(b => b.toString(16).padStart(2, "0"))
        .join("");
}
function _bytes_from_hex(string) {
    let bytes = [];
    string.replace(/../g, function (pair) {
        bytes.push(parseInt(pair, 16));
    });
    return new Uint8Array(bytes).buffer;
}

function _hash(password, salt) {
    return crypto.pbkdf2Sync(
        password,
        salt,
        PBKDF2_ITERATIONS,
        PASSWORD_KEY_LENGTH,
        PBKDF2_DIGEST_ALGO
    );
}

function get_hashed_password(password) {

    const salt_bytes = crypto.randomBytes(SALT_LENGTH);

    // convert password string to byte encoding
    const password_bytes = new TextEncoder().encode(password);

    // calculate hash from password & salt bytes
    const hash_bytes = _hash(password_bytes, salt_bytes);

    // convert to hex and return
    return {
        hashed_password: _hex_from_bytes(hash_bytes),
        salt: _hex_from_bytes(salt_bytes)
    };
}

function check_hashed_password(password_attempt, hashed_password_from_db, salt_from_db) {
    
    // convert everything to bytes
    const password_bytes = new TextEncoder().encode(password_attempt);
    const hash_bytes_from_db = _bytes_from_hex(hashed_password_from_db);
    const salt_bytes_from_db = _bytes_from_hex(salt_from_db);
    
    // calculate hash of attempt with password and salt
    const hash_bytes = _hash(password_bytes, salt_bytes_from_db);

    // compare
    return crypto.timingSafeEqual(hash_bytes, hash_bytes_from_db);
}

function login_required(req, res, next) {
    // kind of like Flask wrapper. Attach to endpoint, gets called before callback
    // router.get('/example', login_required, function (req, res){//code})
    if (req.session.user_id) {
        next();
    } else {
        res.status(401);
        res.send('Access denied');
    }
}

exports.get_hashed_password = get_hashed_password;
exports.check_hashed_password = check_hashed_password;
exports.login_required = login_required;

File: backend-flask/app/auth_utils.py
import hashlib
import hmac
import secrets

SALT_LENGTH = 16
PBKDF2_ITERATIONS = 500000
PASSWORD_KEY_LENGTH = 64
PBKDF2_DIGEST_ALGO = 'sha256'

def _hash(password, salt):
    """Accepts password/salt as byte strings, returns hash as bytes"""
    return hashlib.pbkdf2_hmac(
        PBKDF2_DIGEST_ALGO, 
        password, 
        salt, 
        PBKDF2_ITERATIONS
    )


def get_hashed_password(password):
    """Accepts password as string, returns password hash + salt as bytes"""

    # generate random salt bytes
    salt = secrets.token_bytes(SALT_LENGTH)

    # convert string to byte encoding of that string
    password = password.encode('utf-8')

    # calculate hash from password & salt bytes
    hashed_password = _hash(password, salt)

    # return both, still bytes
    return hashed_password, salt


def check_hashed_password(password_attempt, hashed_password_from_db, salt_from_db):
    """Accepts password attempt as string, hashed_password and salt as hex. Returns bool"""
    
    # convert to byte encoding of string
    password_attempt = password_attempt.encode('utf-8')
    
    # calculate hash of attempt from passwrod/salt bytes
    hashed_attempt = _hash(password_attempt, salt_from_db)

    # compare
    return secrets.compare_digest(hashed_attempt, hashed_password_from_db)
